stbt/android.py

```diff
@@ -0,0 +1,784 @@
+    """
+    CAMERA_720P = 3
+
+
+class AdbDevice(object):
+    """Control an Android device using `ADB`_.
+
+    Default values for each parameter can be specified in your "stbt.conf"
+    config file under the "[android]" section.
+
+    :param string adb_server:
+        The ADB server (that is, the PC connected to the Android device).
+        Defaults to localhost.
+    :param string adb_device:
+        Serial number of the Android device connected to ADB server PC (you can
+        get this by running ``adb devices -l``). If not specified, there must be
+        only one Android device connected. If ``tcpip=True`` this must be the
+        Android device's IP address.
+    :param string adb_binary:
+        The path to the ADB client executable. Defaults to "adb".
+    :param bool tcpip:
+        The ADB server communicates with the Android device via TCP/IP, not
+        USB. This requires that you have enabled TCP/IP ADB access on the
+        device. Defaults to False.
+    :param CoordinateSystem coordinate_system:
+        How to convert the coordinates you give to `AdbDevice.tap` and
+        `AdbDevice.swipe` into the coordinates required by ADB. See
+        `CoordinateSystem` for details. Defaults to ``CAMERA_720P`` on the
+        `Stb-tester CAMERA`_, or ``ADB_NATIVE`` elsewhere.
+    """
+
+    def __init__(self, adb_server=None, adb_device=None, adb_binary=None,
+                 tcpip=False, coordinate_system=None, _config=None):
+
+        if _config is None:
+            _config = stbt._stbt.config._config_init()  # pylint:disable=protected-access
+
+        self.adb_server = adb_server or _get_config(_config, "android",
+                                                    "adb_server", None)
+        self._adb_device = adb_device or _get_config(_config, "android",
+                                                     "adb_device", None)
+        self.adb_binary = adb_binary or _get_config(_config, "android",
+                                                    "adb_binary", "adb")
+        if tcpip is None:
+            try:
+                tcpip = _config.getboolean("android", "tcpip")
+            except ConfigParser.Error:
+                tcpip = False
+        self.tcpip = tcpip
+
+        if coordinate_system is None:
+            name = _get_config("android", "coordinate_system", "ADB_NATIVE")
+            if name not in CoordinateSystem.__members__:  # pylint:disable=no-member
+                raise ValueError(
+                    "Invalid value '%s' for android.coordinate_system in "
+                    "config file. Valid values are %s."
+                    % (name, ", ".join("'%s'" % k for k in
+                                       CoordinateSystem.__members__)))  # pylint:disable=no-member
+            coordinate_system = CoordinateSystem[name]
+        self.coordinate_system = coordinate_system
+
+        if self.tcpip:
+            self._connect(timeout_secs=60)
```

Comment:
I'm not a big fan of this constructor.  It feels a little overloaded.  What about:

```
new_adb_device_usb(adb_device=None, adb_server=None, adb_binary=None, coordinate_system=None)
new_adb_device_tcpip(address, adb_server=None, adb_binary=None, coordinate_system=None)
new_adb_device_from_config(config=None)
```

Hmm, seems a bit much.